main.go

```diff
@@ -14,191 +14,336 @@ import (
+						"task":         stat.Source,
+						"mesos_slave":  host,
+						"framework_id": stat.FrameworkId,
+					},
+				},
+			)
+			memLimitVal.Set(float64(stat.Statistics.MemLimitBytes))
+
+			memRssVal := prometheus.NewGauge(
+				prometheus.GaugeOpts{
+					Namespace: "mesos_task",
+					Name:      "memory_rss_bytes",
+					Help:      "Task memory RSS usage in bytes.",
+					ConstLabels: prometheus.Labels{
+						"task":         stat.Source,
+						"mesos_slave":  host,
+						"framework_id": stat.FrameworkId,
+					},
+				},
+			)
+			memRssVal.Set(float64(stat.Statistics.MemRssBytes))
+
+			outCh <- cpuLimitVal
+			outCh <- cpuSysVal
+			outCh <- cpuUsrVal
+			outCh <- memLimitVal
+			outCh <- memRssVal
+		}
 	}
 }
 
-func processRequests(c *Config) {
-	for _, endpoint := range *c.Endpoints {
-		glog.Infof("Found %s\n", endpoint.Name)
+func (e *PeriodicExporter) lockSlaves(f func() []string) []string {
+	e.slaves.Lock()
+	defer e.slaves.Unlock()
+	return f()
+}
 
-		if len(endpoint.Interval) == 0 {
-			endpoint.Interval = c.Interval
-		}
+func (e *PeriodicExporter) rLockMetrics(f func()) {
+	e.RLock()
+	defer e.RUnlock()
+	f()
+}
 
-		if len(endpoint.TimeoutStr) != 0 {
-			t, err := time.ParseDuration(endpoint.TimeoutStr)
-			if err != nil {
-				glog.Fatalf("Invalid specification of time.Duration: %s\n", err)
-			}
-			endpoint.Timeout = t
-		} else {
-			endpoint.Timeout = 3 * time.Second
+func (e *PeriodicExporter) runEvery(f func(), interval time.Duration) {
+	f()
+	for {
+		select {
+		case <-time.After(interval):
+			f()
 		}
+	}
+}
 
-		go requestRunner(endpoint)
+func (e *PeriodicExporter) setMetrics(ch chan prometheus.Gauge) {
+	metrics := make([]prometheus.Gauge, 0)
+	for metric := range ch {
+		metrics = append(metrics, metric)
 	}
+
+	e.Lock()
+	e.metrics = metrics
+	e.Unlock()
 }
 
-func init() {
-	prometheus.MustRegister(cpuLimitVal)
-	prometheus.MustRegister(cpuSysVal)
-	prometheus.MustRegister(cpuUsrVal)
-	prometheus.MustRegister(memLimitVal)
-	prometheus.MustRegister(memRssVal)
+func (e *PeriodicExporter) scrapeSlaves() {
+	hostnames := e.lockSlaves(func() []string {
+		return e.slaves.hostnames
+	})
+
+	totalHostnames := len(hostnames)
+	glog.V(6).Infof("active slaves: %d", totalHostnames)
+
+	dispatchCh := make(chan string)
+	ch := make(chan prometheus.Gauge)
+	go e.setMetrics(ch)
+
+	poolSize := concurrentFetch
+	if totalHostnames < concurrentFetch {
+		poolSize = totalHostnames
+	}
+
+	glog.V(6).Infof("creating fetch pool of size %d", poolSize)
+
+	var wg sync.WaitGroup
+	wg.Add(poolSize)
+	for i := 0; i < poolSize; i++ {
+		go e.fetch(dispatchCh, ch, &wg)
+	}
+
+	for _, host := range hostnames {
+		dispatchCh <- host
+	}
+	close(dispatchCh)
+
+	wg.Wait()
+	close(ch)
 }
 
-func main() {
-	flag.Parse()
+func (e *PeriodicExporter) updateSlaves() {
+	glog.V(6).Info("discovering slaves...")
+
+	// This will redirect us to the elected mesos master
+	redirectURL := fmt.Sprintf("%s:5050/master/redirect", e.opts.masterURL)
```

Comment:
I think it'd be better / more flexible to let the user specify the port in the command-line flag, since it's part of the URL to the master, and not everyone might use the default port. In the past we've also been bitten by whether user-provided URLs contained a trailing slash or not. Ideally it would allow for both (do something like `strings.TrimRight(masterURL, "/")` at the beginning).